src/ucp/proto/proto_perf.c

```diff
@@ -0,0 +1,411 @@
+            /* Empty range */
+            continue;
+        }
+
+        status = ucp_proto_perf_segment_new(perf, range_start,
+                                            range->max_length, &seg);
+        if (status != UCS_OK) {
+            goto err_destroy;
+        }
+
+        seg->node = range->node;
+        ucs_list_add_tail(&perf->segments, &seg->list);
+
+        ucp_proto_perf_segment_add_func(perf, seg, UCP_PROTO_PERF_FACTOR_SINGLE,
+                                        range->perf[UCP_PROTO_PERF_TYPE_SINGLE],
+                                        NULL);
+        ucp_proto_perf_segment_add_func(perf, seg, UCP_PROTO_PERF_FACTOR_MULTI,
+                                        range->perf[UCP_PROTO_PERF_TYPE_MULTI],
+                                        NULL);
+        ucp_proto_perf_segment_add_func(perf, seg,
+                                        UCP_PROTO_PERF_FACTOR_LOCAL_CPU,
+                                        range->perf[UCP_PROTO_PERF_TYPE_CPU],
+                                        NULL);
+
+        range_start = range->max_length + 1;
+    }
+
+    *perf_p = perf;
+    return UCS_OK;
+
+err_destroy:
+    ucp_proto_perf_destroy(perf);
+err:
+    return status;
+}
+
+ucs_status_t ucp_proto_perf_add_func(ucp_proto_perf_t *perf, size_t start,
+                                     size_t end,
+                                     ucp_proto_perf_factor_id_t factor_id,
+                                     ucs_linear_func_t func,
+                                     ucp_proto_perf_node_t *perf_node)
+{
+    ucp_proto_perf_segment_t *seg, *new_seg;
+    ucs_status_t status;
+    size_t seg_end;
+
+    ucp_proto_perf_check(perf);
+
+    /*                   __________        _________________
+     * perf before:     |__________|      |_________________|
+     *                __________________
+     * range to add: |__________________|
+     *                ______________________________________
+     * perf after:   |__|__________|____|_|_____|__________|
+     */
+    seg = ucs_list_head(&perf->segments, ucp_proto_perf_segment_t, list);
+    while ((&seg->list != &perf->segments) && (start <= end)) {
+        if (start > seg->end) {
+            /* The current segment ends before the requested range */
+            seg = ucs_list_next(&seg->list, ucp_proto_perf_segment_t, list);
+            continue;
+        }
+
+        if (start < seg->start) {
+            /* Insert an empty segment before the current segment */
+            seg_end = ucs_min(end, seg->start - 1);
+            status  = ucp_proto_perf_segment_new(perf, start, seg_end, &new_seg);
+            if (status != UCS_OK) {
+                return status;
+            }
+
+            ucs_list_insert_before(&seg->list, &new_seg->list);
+            seg = new_seg;
+        } else {
+            /* Split the first affected segment */
+            if (start > seg->start) {
+                ucs_assertv(start > 0, "perf=%p start=%zu", perf, start);
```

Comment:
`start` and `seg->start` are both `size_t`, so `start` can never be 0 if it is greater than `seg->start`